### PR TITLE
Add popup with origin tab url

### LINKED ISSUE #2: More control over what happens if extension button is clicked
I want to propose this:

Extension popup window with "origin" url and "go to" button. For the case if a user wants to just know what was the "origin" url but not to open it. 

What exactly happens when a user clicks extension button is controlled by setting with these options:

- open and go to origin tab (default behavior)
- show popup if origin tab is closed
- always show popup

If you like this I will do a PR.

### PR DESCRIPTION
**Pull request by T1mL3arn**
closes #2

## Changes:

- extension settings page to enable popup (there are 3 options: open tab, show popup, go to tab if it is already open)
- popup itself
- migrate to manifest v3 (for now this extension works only in Firefox)

## How does it look

image

## Notes

It is draft to receive feedback, in particular about migration to v3.

### CODE REVIEW

**Comment by pelmers** [comment]
Wow, nice work! I just started a vacation today and left the computer at home. I’ll try to remember to review but ping me if I have not responded by November 1. 
Quick note on skimming: I saw  that the main commit also has a lot of white space changes, ideally could you patch those into the fix indent commit?
Also thanks for addressing manifest v3. I vaguely knew that was coming up but never got around to acting on it.

**Comment by T1mL3arn** [comment]
Some todos:

- Set min chrome/ff version in manifest
- Decide if FF should use `module` for bg script.  FF started to [support](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background#browser_compatibility) `type=module` for bg scripts only recently, so to support older FF versions webpack should be introduced (or smth).

**Comment by pelmers** [comment]
Hi, I see you're still working on this, I returned from my trip so just let me know if you want some feedback or code review.

**Comment by T1mL3arn** [comment]
I'd say I implemented everything I wanted to, so patch is ready for review.

File: background.js [hunk 1]
```diff
@@ -1,8 +1,16 @@
+import { actionType, getActionType, actionTypeKey, TAB_ORIGIN_DATA_MSG } from '/lib.js';
+import * as lib from '/lib.js';
+
 // TODO: port to browser api
 const api = chrome;
 
 const activeTabKey = "activeTab";
 
+let lastTabOriginData = {}
```
Comment (pelmers) [inline background.js:9]: can you add a comment on the purpose of this variable?

Comment (pelmers) [inline background.js:17]: can you leave a comment on the purpose of this variable?

File: background.js [hunk 2]
```diff
@@ -11,31 +19,94 @@ function last(array) {
...
+ * current action (open tab, show popup, etc.)
+ */
+function extensionButtonClick(tab) {
+
+    lastTabOriginData = {}
+
+    const tabId = tab.id.toString();
+
+    // TODO since bg script is not persistent
+    // it is possible that `currentActionType` will be `undefined`
```
Comment (pelmers) [inline background.js:49]: do you happen to know what conditions trigger this?

File: background.js [hunk 3]
```diff
@@ -11,31 +19,94 @@ function last(array) {
...
+                    const tabOriginData = originTabIsOpen ? 
+                        { tabId: originTabId, windowId, url } : { url, currentTabIndex, currentTabId: tab.id }
+
+                    if (currentActionType === actionType.SHOW_POPUP) {
+
+                        lastTabOriginData = tabOriginData
+                        
+                    } else if (currentActionType === actionType.GO_TO_TAB_IF_OPEN) {
+
+                        if (originTabIsOpen)
```
Comment (pelmers) [inline background.js]: let's add braces for consistency

File: background.js [hunk 4]
```diff
@@ -11,31 +19,94 @@ function last(array) {
...
+                            lib.focusTab(originTabId, windowId)
+                        else {
+                            lastTabOriginData = tabOriginData
+                        }
+
+                    } else {
+                        // otherwise it is default "open tab" action
+
+                        if (originTabIsOpen)
```
Comment (pelmers) [inline background.js]: {}

File: lib.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+export function focusTab(tabId, windowId) {
+	chrome.tabs.update(tabId, { active: true });
+	chrome.windows.update(windowId, { focused: true });
+}
+
+export function createTab(url, index, historyStack) {
+	// createTab() can reject when you try to open
+	// privileged urls like "about:config",
+	// see https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/Tabs/create#url
```
Comment (pelmers) [inline lib.js]: thanks for adding this comment

File: lib.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
+export const actionTypeKey = 'extension-action-type'
```
Comment (pelmers) [inline lib.js:1]: change tab to space indentation in this file

**Review by pelmers — CHANGES_REQUESTED** [review]
really sorry about the late response. I have tested out your patch on Chrome and Firefox. It is indeed a little bothersome that the two browsers seem to require each their own copy of the manifest file. In Chrome I was getting errors with the "openPopup" API, do you know if that's available yet?
In Firefox all three of the behaviors seemed to work well. Do you know from which version does Firefox support module scripts?
I left a couple of comments but overall the code is pretty reasonable.

Thanks for your work! If you can fix up my minor comments then I will merge and publish a new version.

File: background.js (hunk 2 above)
Comment (T1mL3arn) [reply background.js:49]: In which circumstance bg script is terminated? I don't know. If you are talking about `currentActionType` - I don't remember, this comment looks to me as a complete mystery right now :sweat_smile: . Maybe I got a couple of null-ref errors. I assume the comment it is not relevant anymore,

**Comment by T1mL3arn** [comment]
>Do you know from which version does Firefox support module scripts?

I place a direct link to compat table in some post above, according to the table it is 112.

**Comment by pelmers** [comment]
thank you for your work! I will test it a bit more and then release to the extension stores

**Pull request merged**